app/models/GoodsRecordsPagination.scala

```diff
@@ -32,47 +31,67 @@ case class GoodsRecordsPagination(
 
 object GoodsRecordsPagination {
   implicit val format: OFormat[GoodsRecordsPagination] = Json.format[GoodsRecordsPagination]
+  private val defaultRecord                            = 0
 
-  private def getPageSize(totalRecords: Int, recordsSize: Int, totalPages: Int): Int =
-    (totalRecords - recordsSize) / (totalPages - 1)
+  def getFirstRecord(pagination: GoodsRecordsPagination, pageSize: Int): Int =
+    if (
+      pagination.totalRecords == 0 || pagination.currentPage > pagination.totalPages || (pagination.currentPage > 1 && pageSize >= pagination.totalRecords)
+    ) {
+      defaultRecord
+    } else {
+      (
+        (pagination.currentPage - 1) * pageSize
+      ) + 1
+    }
 
-  def getFirstRecord(goodsRecordResponse: GetRecordsResponse): Int =
-    getFirstRecordPos(
-      goodsRecordResponse.pagination.totalRecords,
-      goodsRecordResponse.goodsItemRecords.size,
-      goodsRecordResponse.pagination.totalPages,
-      goodsRecordResponse.pagination.currentPage
-    ) + 1
+  def getLastRecord(firstRecord: Int, len: Int): Int =
+    if (firstRecord == 0) {
+      defaultRecord
+    } else {
+      len + firstRecord - 1
+    }
 
-  private def getFirstRecordPos(totalRecords: Int, recordsSize: Int, totalPages: Int, currentPage: Int): Int =
-    (currentPage - 1) * getPageSize(totalRecords, recordsSize, totalPages)
+  def getPagination(currentPage: Int, totalPages: Int): Pagination =
+    if (currentPage < 1 || totalPages < 1 || currentPage > totalPages) {
+      Pagination(None, None, None)
+    } else {
 
-  def getLastRecord(goodsRecordResponse: GetRecordsResponse): Int =
-    goodsRecordResponse.goodsItemRecords.size + getFirstRecordPos(
-      goodsRecordResponse.pagination.totalRecords,
-      goodsRecordResponse.goodsItemRecords.size,
-      goodsRecordResponse.pagination.totalPages,
-      goodsRecordResponse.pagination.currentPage
-    )
-
-  def getPagination(pagination: GoodsRecordsPagination): Pagination =
-    Pagination(
-      items = Some((0 until pagination.totalPages).map { index =>
-        PaginationItem(
-          number = Some((index + 1).toString()),
-          current = Some(pagination.currentPage == (index + 1)),
-          href = routes.GoodsRecordsController.onPageLoad(index + 1).url
-        )
-      }),
-      previous = if (pagination.currentPage == 1) {
-        None
+      val start = if (currentPage <= 2) {
```

Comment:
not sure if doing like Max(1, currentPage-2) might be more comprehensible. This is good though, just took me some playing to figure out what everything meant and why -2 etc was floating around. Impressed by how much hassle it is just to get that little page number thing at the bottom of the list!

Reply:
i'll just leave it as it is